### ISSUE TITLE
Character encoding issues in Postman (master issue)

### ISSUE THREAD

**Issue by P1llus** [a85/POSTMan-Chrome-Extension#4555]
Hello everyone!

Several request has been made in terms of encoding errors on get requests on Postman requests, some are old (and resolved in like v4-5), but it has come back again.

Postman in general is of course a open source tool, where people respond and fix certain issues based on feedback on their free time and i still fully respect that, but i have seen that this request has been generally ignored while both older and newer requests are swiftly responded to, both my own and others, and this is my last effort to see if we can at least get a respond if this is being looking into or not, as i love to use postman for everything as i have always done and always will unless it does not work.

Here is a few examples, both closed and open:

#2538, #2468, #2508, #2510, #3329, #2517, #4507, #4444.

I hate tagging for response, but for once i hope the person will accept it, just for a response, not saying you need to fix it or anything.. @prashantagarwal

**Comment by SamvelRaja** [comment]
@P1llus  First thanks for taking your time and bringing all the issues to one place. 

And yeah, we are on the roadmap of features where we missed to reply few issues. Let me take my time off and reply everything and also answer your issues here.

As you mentioned, yes postman is a very generic tool and for URL encoding we are following the [RFC 3986
](https://www.ietf.org/rfc/rfc3986.txt) and encoding algorithm specified in [what-wg format](https://url.spec.whatwg.org/#parsers) based on that we created a library [postman-url-encoder](https://github.com/postmanlabs/postman-url-encoder), which is been used in our request sending engine - [Postman-runtime](https://github.com/postmanlabs/postman-runtime) 

So, any implementation we come up with, we stick to the spec.

#2538 and #2468 - These issues were people started seeing the reserved delimiters got encoded, so we decided to create the above-mentioned library and fixed it.

The main agenda of these two issues are, '+' sign in the URL param got encoded automatically, which is against the RFC and most servers consider, '+' as 'space' or '+' itself when it receives it.

So, we stick to the RFC and provided the general solution of '+' will not be encoded along with its other delimiters. 

#4507 #4444 - These issues contradicts the fixes made for the above-mentioned issues, where it specified the '+' needs to be encoded in the URL param before sending to the server.

The alternate way to solve this is by sending `%2B` instead of `+`. 

But, We are not letting you down like this, we will find a way to solve this problem also in more generic way by providing config or so.

#2517 - Where people claim that even when they encode '+' to '%2B' it is reaching the server as 'space', It was not reproducible and the logs also says postman is not tampering the already encoded data. I will add the appropriate recordings on that issue as well.

#3329 - Yeah, this is a very interesting issue, where people want to encode, '+' to %20' instead of '%2B', since both are allowed, we need to stick to one. We will anyway think of a generic solution of providing this also as a config to satisfy both sides of the coin.

#2508 - It is a UI issue where the # is behaving differently, we will push this to our roadmap and accommodate as soon as possible.

#2510 - This is on the request body, which is out of this issues context. But I will reply in the issue regarding that.

And, I really appreciate your patience in this. we are actively working on solving the problems for the whole community.

And our URL encoder is open sourced and we accept PR if you find that we are moving away from standard :)

**Comment by P1llus** [comment]
@SamvelRaja Thank you very much for your kind response! :) The linked issues itself was just to point out the general problem with encoding that people seemed to have, and even though i did search through both open and closed issues i feel that with this response i should have done a better job both explaining and referring to issues. So let me try again! **Sorry for the long text, feel free to skip to the bottom, a small example included in the middle though.**

**I might come off as a bit of an ass, but i promise that non of this text is directed towards anyone in any bad manner or intent**

I use Postman heavily for testing, Collections can come up to 2000+ different requests and more, all made in a very specific way, to ensure that almost all code can be reused, where the Collection layer handles authentications, sessions and user specific information that is reusable throughout testing, using pre-request scripts.

Each subfolder then has it's own pre-request and test scripts, to automatically create/modify/delete resources based on which subfolder you are in, to be able to reuse 90% of my code in all subfolders.

A very generic example would be to only fill out username + password variables in the collection, do an API call to a edit blog page without any preparations, it then notices you don't have a session and creates one for you, then stores all user information in variables for reuse, it then notices you don't have any blog post to edit, so it creates that as well + store relevant id's and so on, which unfortunately makes it quite hard to automate the process of changing out the + characters, and cannot be done manually.

I have spent countless of hours implementing this, and i feel by no means that gives me any privilege to demand any changes of any sort, just so that is pointed out, but since everything i do is automated, and both session id's, resource id's, user information etc etc CAN include a plus sign, and it all is broken by a single non URL encoded character i won't lie and say i don't feel a bit "meh".

Of course a response could also mention "why do you even transfer that data in GET?" This was just an example, most of my API requests are unfortunately towards closed sourced legacy vendors, but it at least gives me an easy way of explaining it.

Okay, with that out of the way, let me see if i can give a proper response as well!

#2538 and #2468 It might have been a weird two issues to bring up, as you say they totally contradict what i created my issue for, but it does highlight the issue of how an implementation is decided, and what type of implementation has the least breaking impact. I feel that so many services out there requires your full URL to be compiled into the correct format before sending, and just removing that functionality might have broken more than it could fix? Or maybe i sound crazy?

#4507 #4444 These issues did indeed point out the problem with encoding, but your response made me a bit confused, if you would care to elaborate on this. First you mentioned that you created the url-encoder to be used in the Postman Runtime application, but then you pointed out that the solution would be to encode the request?
On 2.2.  Reserved Characters in the RFC, it states that:
>URI producing applications should percent-encode data octets that
   correspond to characters in the reserved set unless these characters
   are specifically allowed by the URI scheme to represent data in that
   component.  If a reserved character is found in a URI component and
   no delimiting role is known for that character, then it must be
   interpreted as representing the data octet corresponding to that
   character's encoding in US-ASCII.

So i can see where the issue arise, when should you encode, and when should you provide encoding to represent US-ASCII.

#3329 When/How do you decide if a restricted character should be encoded or not? The RFC does not mention that restricted characters should be ignored, but rather decided by the application if the URI requires encoding to work.

@SamvelRaja In the end, all i was looking for of course, was hopefully a way to resolve my issue, so i can continue to fully utilize Postman again! :) Feel free to just respond to smaller parts of this, to save up more time for more awesome Postman ;)

I will bookmark the URL Encoder and hopefully i get some time to take a look soon enough!

**Comment by SamvelRaja** [comment]
Since this issue has the most bundled context of the encoding issues people facing, I am referencing this and closing all the other ones and will take a deep look and will come up with the more generic solution sooner.

**Comment by oliverjanik** [comment]
Wow, what a mess. Why is it hard to follow the spec?

**Comment by jmini** [comment]
Here is my expectation:

In the URL field: see the request as it will be sent (special chars needs to be encoded)
In the param field: see the value in clear text (special chars are decoded)

<img width="730" alt="expected" src="https://user-images.githubusercontent.com/1222165/44251377-6992f480-a1f8-11e8-8a1b-42124bb26952.png">


---

Other example:

If you would have a `+` in the url field of postman:
```
?name=jbr+email
```
The value in the parameter field should be: `jbr email`, because this is what the server will get.

**Comment by kishoretangudu** [comment]
is the + symbol issue got resolved? can anybody respond me pls. I am using postman 5.5.4. I still see the same issue.

**Comment by EverestJoo** [comment]
yes, postman with old version by using the chrome app has the feature which can do the urlencoding for the get params and post params. However the newest version, v 6.7.2, can not automatically do the urlencoding for the request parameters, I have to do this manually. 
Hope to fix this issue soon!

**Comment by erodewald** [comment]
Still can't use `+` in query params in `v7.0.6` macOS.

**Comment by dumbdonkey** [comment]
this is a problem hard to found, I've debugged my server then found postman didn't encode '+' and my server treats it as space then everything became unclearly. Hope your fixing

**Comment by mcurrao** [comment]
Hey there. Faced same issue here now. 
First, I would like to say I completely agree with @jmini 's expectation on encoding ( [Related comment on same thread](https://github.com/postmanlabs/postman-app-support/issues/4555#issuecomment-413772061) ).

I had to debug down to Spring's inner `ExtendedServletRequestDataBinder` to realize I was receiving the date as '2017-08-31T00:00:00.000 05:00' instead of '2017-08-31T00:00:00.000+05:00', and that it was not Spring's date conversion system's fault. Good news: that works.

Which has in turn lead me to realize that the URL field of Postman does not show the real URL (as I purposedly added an empty space in the "Params" and it reflected as a space in the URL). I would have expected **the URL input to show the fully encoded URL**, and **the "Params" input to show unencoded params**. This is (at least for me) the best solution. You get RFC-compliant URLs you are able to drop on/from any browser, and at the same time you get unencoded, easy to read parameters in their own field.

Really hope you can take a look into this, as it can lead to serious unexpected behaviours (my request would work with negative-offset dates, for example). Thanks.

**Comment by gigamac** [comment]
Hi
Getting same issue in postman v6.7.4 with Japanese characters in a GET request - is there a solution or not? I am getting 404 but when I copy and paste the (unencoded) url into chrome or Firefox I get the correct result. Safari returns 404 too btw

**Comment by TenLetters88** [comment]
Hi,

Any update in GET calls with query parameters with '+' in the params? It's still not working, regardless of using '+' sign or %2B.

Thank you!

**Comment by RetroAsmDev** [comment]
The "+" symbol should is still not encoded in URL queries as "%2B" in Postman ver. Version 7.11.0 on win32 6.1.7601 / x64. I have to encode it manually in Pre-test scripts.

**Comment by EmirhanAksoy** [comment]
still same

"+" character came as string empty  in query string

**Comment by kadlugan** [comment]
I recently came across an issue where white spaces in xml payloads were being converted from %20 to %C2%A0 when you request **pretty** response text payloads.  This occurred using Windows 10 env, Excel 2016, postman v7.15.0

- You can test by copying whitespace text from the **raw** and the **pretty** versions of xml payloads and pasting into an excel cell.  The text has to contain a space e.g. "2001 A Space Odyssey"
- Then by running =ENCODEURL(text) and it shows %C2%A0 as the whitespace values. 
- Heres a link to what that character is http://www.fileformat.info/info/unicode/char/00a0/index.htm
- I was pasting the values back into XML dynamic queries and using postmans EncodeURIComponent and seeing those same spaces that were %20 in the raw become %C2%A0 values. 
- The net results was my comparison tests were failing because a %20 whitespace does not equal a %C2%A0 whitespace.

Something to be aware of.

**Comment by shamasis** [comment]
Hello everyone in this thread! I think we’ve made some progress on this aspect that would interest you. We’ve updated the URL processor with a new one in our latest Canary release. Would love to gather your feedback on the same. More information on what we did is at: https://github.com/postmanlabs/postman-app-support/issues/8154

**Comment by codenirvana** [comment]
Hey everyone 👋
It took us a while but, we finally fixed most of the issues related to URL encoding. Check #8154 for more details.

The new URL processor is available in the latest version of the Postman app (v7.20.0) by enabling the "Use next generation URL processing" setting.
<img width="1552" alt="postman-settings" src="https://user-images.githubusercontent.com/4865836/76351543-a0c57b00-6333-11ea-8463-36b23d159456.png">

Regarding the discussion around encoding the `+` character or converting the whitespaces to `+` I suggesting opening a [new issue](https://github.com/postmanlabs/postman-app-support/issues/new/choose) (if the issue persists) with all the required details because this thread is referenced for every other URL encoding related issues which made this thread very confusing.

The new URL processor follows the [WHATWG (URL) standards](https://url.spec.whatwg.org/) like the majority of the browsers and other clients and has a consistent URL encoding behavior as follows:

URL Component | Characters to Encode
-- | --
Path | `SPACE`, `"`, `<`, `>`, `#`, `?`, `{`, `}`, <code>`</code>
Query | `SPACE`, `"`, `#`, `&`, `'`, `<`, `=`, `>` 
User info | `SPACE`, `"`, `<`, `>`, `#`, `?`, `{`, `}`, `/`, `:`, `;`, `=`, `@`, `[`, `\`, `]`, `^`, `|`, <code>`</code>

Feel free to leave your feedback at #8154. Closing this thread now.

**Comment by fabiendeborde** [comment]
Hello, I have an issue with Postman encoding and hope this is the right place to post about it.
I have some Japanese text in my OpenAPI document, and I recently synced it with a GitHub repository, but the Japanese text comes out wrongly encoded and messes with the schema validity.

For example, this code
```
openapi: 3.0.0
info:
  title: My API
  version: 2.0.0
  description: 'My app 管理画面専用API'
```

will be saved as this
```
openapi: 3.0.0
info:
  title: My API
  version: 2.0.0
  description: 'My app �;b(API'
```
I don't know if it is needed but I am using Postman 7.22.1 on latest Mac version (10.15.4).
Is there any way to fix this?

**Comment by fabiendeborde** [comment]
@codenirvana 
I have no clue if I should comment on this closed issue or write a new one, as it seems to me that this is the same problem. Can you please hint me into the right direction?

**Comment by codenirvana** [comment]
@FabienDeborde This issue is related to encoding issues in the request URL.
I suggest you file a separate issue for this particular problem.

**Comment by aliraza-lakhani** [comment]
I'm using postman chrome extension, facing issue when sending + it gets converted into space,
Is there any solution to force + sign in request parameter?

**Comment by arlemi** [comment]
@arlmath We do not support the Chrome extension anymore, please use the app instead: https://www.postman.com/downloads/

**Comment by tartale** [comment]
@arlemi I see the `+` issue in the app as well (`Version 7.27.1 (7.27.1)`)

**Comment by codenirvana** [comment]
@arlmath @tartale as mentioned in https://github.com/postmanlabs/postman-app-support/issues/8626#issuecomment-642485743 Postman don't convert/encode the `+` sign but a lot of query processors used in the requesting server do treats `+` as `space`.

So, we are planning to have a per-request level setting to automatically encode the `+`s. Follow #8626 for more updates.

**Comment by gyana12** [comment]
i am using latest postman version, i am facing issue with respect to ÷ division char. If i am sending empty request {} body then the encoding getting changed to Ã·

**Comment by mhtmalpani** [comment]
A hack is to make use of the Pre-request Script in Postman.

Paste the following in the Pre-request Script section. This will transform all our params in the Request Url to be encoded.
Thus all the `+` sign would be encoded to `%2B` 

```javascript
var querycount = pm.request.url.query.count();
for(let i = 0; i < querycount; i++) {
  pm.request.url.query.idx(i).value = encodeURIComponent(pm.request.url.query.idx(i).value);
}
```
